File: store/src/main/java/com/min/store/product/service/ProductService.java
package com.min.store.product.service;


import com.min.store.product.domain.*;
import com.min.store.product.entity.ProductEntity;
import com.min.store.product.repository.ProductRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import javax.persistence.EntityNotFoundException;
import java.util.List;

@RequiredArgsConstructor
@Service
@Transactional(readOnly = true)
public class ProductService {

    private final ProductRepository productRepository;

    public List<ProductDomain> getProducts(){
        List<ProductEntity> productEntities = productRepository.findAll();

        return productEntities.stream()
                .map(ProductEntity::toDomain)
                .toList();
    }

    public ProductDomain getById(Long id){
        ProductEntity product = productRepository.findById(id)
                .orElseThrow(EntityNotFoundException::new);

        return product.toDomain();
    }

    public ProductDomainKotlin getByIdKotlin(Long id){
        ProductEntity product = productRepository.findById(id)
                .orElseThrow(EntityNotFoundException::new);

        //return product.toDomain();
        return product.toDomainKotlin();
    }

    @Transactional
    public ProductDomain register(ProductCreateDomainKotlin productCreateDomain){

        ProductDomain productDomain = ProductDomain.from(productCreateDomain.getName(), productCreateDomain.getPrice());

        duplicateProductCheck(productDomain);

        ProductEntity productEntity = ProductEntity.from(productDomain);

        return productRepository.save(productEntity)
                .toDomain();
    }

    private void duplicateProductCheck(ProductDomain productDomain){
        boolean isProduct = productRepository.existsByName(productDomain.getName());
        if(isProduct) throw new IllegalStateException("이미 등록된 상품입니다.");
    }

    @Transactional
    public ProductDomain update(Long id, ProductUpdateDomain productUpdateDomain){
        ProductDomain productDomain = getById(id);
        productDomain.update(productUpdateDomain);

        return productRepository.save(ProductEntity.from(productDomain))
                .toDomain();
    }
}

File: store/src/main/java/com/min/store/product/service/ProductService.kt
package com.min.store.product.service

import com.min.store.product.domain.ProductCreateDomain
import com.min.store.product.domain.ProductDomain
import com.min.store.product.domain.ProductUpdateDomain
import com.min.store.product.entity.ProductEntity
import com.min.store.product.repository.ProductRepository
import lombok.RequiredArgsConstructor
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
open class ProductService (private val productRepository: ProductRepository){

    fun getProducts() : List<ProductDomain> {
        val productEntitys : List<ProductEntity> = productRepository.findAll()

        return productEntitys.map { it.toDomain() }
            .toList()
    }

    fun getById(id : Long) : ProductDomain {
        val product: ProductEntity = productRepository.findByIdOrNull(id) ?: throw NoSuchElementException()

        return product.toDomain()
    }

    fun getByName(name : String) : ProductDomain {
        val product: ProductEntity = productRepository.findByName(name) ?: throw NoSuchElementException()

        return product.toDomain()
    }

    fun register(productCreateDomain:ProductCreateDomain) : ProductDomain {
        val productDomain = ProductDomain.from(productCreateDomain.name, productCreateDomain.price)

        duplicateProductCheck(productDomain);

        val productEntity:ProductEntity = ProductEntity.from(productDomain)

        return productRepository.save(productEntity)
            .toDomain()
    }

    private fun duplicateProductCheck(productDomain : ProductDomain) {
        val isProduct = productRepository.existsByName(productDomain.name)
        if(isProduct) throw IllegalStateException("이미 등록된 상품입니다.")
    }

    @Transactional
    open fun update(id: Long, productUpdateDomain: ProductUpdateDomain) : ProductDomain {
        val productDomain:ProductDomain = getById(id)
        productDomain.update(productUpdateDomain)

        return productRepository.save(ProductEntity.from(productDomain))
            .toDomain()
    }


}
